pilot/model/service.go

```diff
@@ -453,3 +455,87 @@ func ParseLabelsString(s string) Labels {
 	}
 	return tag
 }
+
+func buildUniqueSet(values []string) map[string]bool {
+	out := map[string]bool{}
+	for _, val := range values {
+		out[val] = true
+	}
+	return out
+}
+
+func setEquals(first, second map[string]bool) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	for firstVal := range first {
+		_, found := second[firstVal]
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+// Equals compares one list of ports with another
+func (ports *PortList) Equals(other *PortList) bool {
+	if len(*ports) != len(*other) {
+		return false
+	}
+	plMap := map[int]*Port{}
+	for _, p := range *ports {
+		plMap[p.Port] = p
+	}
+	otherMap := map[int]*Port{}
+	for _, p := range *other {
+		otherMap[p.Port] = p
+	}
+	for portNum, p := range plMap {
+		o, found := otherMap[portNum]
+		if !found {
+			return false
+		}
+		if p.Name != o.Name ||
+			p.Protocol != o.Protocol ||
+			p.AuthenticationPolicy != o.AuthenticationPolicy {
+			return false
+		}
+	}
+	return true
+}
+
+// Equals compares services based on only what's needed for xDS to work.
+// This method assumes service names are the same and compares
+// other relevant fields.
+func (s *Service) Equals(other *Service) bool {
```

Comment:
You almost never need to implement an `Equals` method in golang, you typically want to use `reflect.DeepEquals`.